text/0000-cargo-prepublish.md

```diff
@@ -0,0 +1,422 @@
+  uses of a particular existing crate version to new source for the crate, even
+  if there are breaking changes involved. The feature is designed for surgical
+  patching of specific dependency versions.
+
+Prepublication dependencies add another tool to this arsenal, with just the
+right amount of dependency unification: the precise amount you'd get after
+publication to crates.io.
+
+# Detailed design
+[design]: #detailed-design
+
+The design itself is relatively straightforward. The Cargo.toml file will
+support a new section for augmenting a source of crates:
+
+```toml
+[augment.crates-io]
+xml-rs = { path = "path/to/fork" }
+```
+
+The listed dependencies have the same syntax as the normal `[dependencies]`
+section, but they must all come form a different source than the source being
+augmented. For example you can't augment crates.io with other crates from
+crates.io! Cargo will load the crates and extract the version information for
+each dependency's name, supplementing the source specified with the version it
+finds. If the same name/version pair *already* exists in the source being
+augmented, then this will act just like `[replace]`, replacing its source with
+the one specified in the `[augment]` section.
+
+Like `[replace]`, the `[augment]` section is only taken into account for the
+root crate (or workspace root); allowing it to accumulate anywhere in the crate
+dependency graph creates intractable problems for dependency resolution.
+
+The sub-table of `[augment]` (where `crates-io` is used above) is used to
+specify the source that's being augmented. Cargo will know ahead of time one
+identifier, literally `crates-io`, but otherwise this field will currently be
+interpreted as a URL of a source. The name `crates-io` will correspond to the
+crates.io index, and other urls, such as git repositories, may also be specified
+for augmentation. Eventually it's intended we'll grow support for multiple
+registries here with their own identifiers, but for now just literally
+`crates-io` and other URLs are allowed.
+
+## Examples
+
+It's easiest to see how the feature works by looking at a few examples.
+
+Let's imagine that `xml-rs` is currently at version `0.9.1` on crates.io, and we
+have the following dependency setup:
+
+- Crate `foo` lists dependency `xml-rs = "0.9.0"`
+- Crate `bar` lists dependency `xml-rs = "0.9.1"`
+- Crate `baz` lists dependency `xml-rs = "0.8.0"`
+- Crate `servo` has `foo`, `bar` and `baz` as dependencies.
+
+With this setup, the dependency graph for Servo will contain *two* versions of
+`xml-rs`: `0.9.1` and `0.8.0`. That's because minor versions are coalesced;
+`0.9.1` is considered a minor release against `0.9.0`, while `0.9.0` and `0.8.0`
+are incompatible.
+
+### Scenario: augmenting with a bugfix
+
+Let's say that while developing `foo` we've got a lock file pointing to `xml-rs`
+`0.8.0`, and we found the `0.8.0` branch of `xml-rs` that hasn't been touched
```

Comment:
Something seems wrong here? We've said that foo depends on 0.9.0, so there's no way we can see 0.8.0 in the lockfile.